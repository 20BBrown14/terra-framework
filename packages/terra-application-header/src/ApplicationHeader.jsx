import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import 'terra-base/lib/baseStyles';

import AppDelegate from 'terra-app-delegate';
import IconMenu from 'terra-icon/lib/icon/IconMenu';
import Button from 'terra-button';

import styles from './ApplicationHeader.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * The AppDelegate instance provided by the containing component. If present, its properties will propagate to the children components.
   * */
  app: AppDelegate.propType,
  /**
   * Content element to be placed within the fill area of the header.
   * */
  content: PropTypes.element,
  /**
   * Logo element to be placed at the start of the header.
   * */
  logo: PropTypes.element,
  /**
   * Layout configuration to handle size and toggle functionality.
   * */
  layoutConfig: PropTypes.shape({
    size: PropTypes.string,
    toggleMenu: PropTypes.func,
    menuIsOpen: PropTypes.bool,
  }),
  /**
   * Utility element to be placed at the end of the header.
   * */
  utility: PropTypes.element,
  /**
   * Widget element to be placed beforethe end of the header.
   * */
  widget: PropTypes.element,
};

const defaultProps = {
  layoutConfig: {
    menuIsOpen: false,
  },
};

const appendPropsToElement = (app, size, element) => React.cloneElement(element, { app, size });

const ApplicationHeader = ({
  app,
  content,
  layoutConfig,
  logo,
  utility,
  widget,
  ...customProps
}) => {
  const isCompact = layoutConfig.size === 'tiny' || layoutConfig.size === 'small';

  const headerClassNames = cx([
    'header',
    'fill',
    customProps.className,
  ]);

  let logoElement;
  if (logo) {
    const clonedElement = appendPropsToElement(app, layoutConfig.size, logo);
    logoElement = <div className={cx(['fit', 'start', 'logo'])}>{clonedElement}</div>;
  }

  let contentElement;
  if (content) {
    const clonedElement = appendPropsToElement(app, layoutConfig.size, content);
    contentElement = <div className={cx('fill', 'content')}>{clonedElement}</div>;
  }

  let widgetElement;
  if (widget) {
    const clonedElement = appendPropsToElement(app, layoutConfig.size, widget);
    widgetElement = <div className={cx(['fit', 'end', 'widget'])}>{clonedElement}</div>;
  }

  let utilityElement;
  if (utility) {
    const clonedElement = appendPropsToElement(app, layoutConfig.size, utility);
    utilityElement = <div className={cx(['fit', 'end', 'utility'])}>{clonedElement}</div>;
  }

  let headerToggle;
  if (layoutConfig.toggleMenu && isCompact) {
    headerToggle = (
      <div className={cx(['fit', 'header-toggle'])}>
        <Button className={cx('toggle-button')} variant="link" icon={<IconMenu />} onClick={layoutConfig.toggleMenu} />
      </div>
    );
  }

  let headerBody;
  if (logoElement || contentElement || utilityElement) {
    headerBody = (
      <div className={cx(['fill', 'header-body'])}>
        {logoElement}
        <div className={cx(['fill'])}>
          {contentElement}
          {widgetElement}
        </div>
        {utilityElement}
      </div>
    );
  }

  return (
    <div {...customProps} className={headerClassNames}>
      {headerToggle}
      {headerBody}
    </div>
  );
};

ApplicationHeader.propTypes = propTypes;
ApplicationHeader.defaultProps = defaultProps;

export default ApplicationHeader;
