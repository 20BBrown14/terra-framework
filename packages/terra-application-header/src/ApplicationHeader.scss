/* stylelint-disable selector-class-pattern */
:local {
  .fit {
    flex: 0 0 auto;
  }

  .fill {
    flex: 1 1 auto;
    justify-content: space-between;
    min-width: 0;
    width: 100%;
  }

  .fit,
  .fill { 
    align-items: center;
    display: flex;
    height: 100%;
    position: relative;
  }

  .header {
    background-color: #2481ca;
    height: 2.5rem;
  }

  .header-toggle {
    border-right: 0.07143rem solid rgba(255, 255, 255, 0.17);
    box-shadow: inset -0.07143rem 0 0 0 rgba(0, 0, 0, 0.17);
    width: 3.0714em;
  }

  .toggle-button {
    color: #fff;
    width: 100%;
  }

  .toggle-button:hover {
    color: #fff;
  }

  .start {
    justify-content: flex-start;
  }

  .end {
    justify-content: flex-end;
  }

  .logo {
    min-width: 18.571rem;
  }

  .logo,
  .utility {
    max-width: 40%;
  }

  .middle {
    max-width: 100%;
  }
}
