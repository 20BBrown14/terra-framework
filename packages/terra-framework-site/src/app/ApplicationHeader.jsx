import React from 'react';
import PropTypes from 'prop-types';
import Image from 'terra-image';
import IconSettings from 'terra-icon/lib/icon/IconSettings';
import Header from 'terra-application-header';
import NavTabs from './common/nav-tabs/NavTabs';

const propTypes = {
  layoutConfig: PropTypes.object,
  navigation: PropTypes.object,

  locale: PropTypes.string,
  onLocaleChange: PropTypes.func,

  dir: PropTypes.string,
  onDirChange: PropTypes.func,

  theme: PropTypes.string,
  onThemeChange: PropTypes.func,
};

class ApplicationHeader extends React.Component {
  render() {
    const isCompactHeader = (this.props.layoutConfig.size === 'tiny' || this.props.layoutConfig.size === 'small');

    const bidiContent = (
      <Header.Utility.ItemGroup key="site-bidi" isSelectable dir="ltr" size="medium" onChange={this.props.onDirChange}>
        <Header.Utility.Item id="ltr" text="ltr" key="ltr" isSelected={this.props.dir === 'ltr'} />
        <Header.Utility.Item id="rtl" text="rtl" key="rtl" isSelected={this.props.dir === 'rtl'} />
      </Header.Utility.ItemGroup>
    );

    const localeContent = (
      <Header.Utility.Item
        text={`Locale: ${this.props.locale}`}
        key="locale"
        subMenuItems={[
          <Header.Utility.ItemGroup isSelectable key="local-options" onChange={this.props.onLocaleChange} >
            <Header.Utility.Item id="en" text="en" key="en" isSelected={this.props.locale === 'en'} />
            <Header.Utility.Item id="en-GB" text="en-GB" key="en-GB" isSelected={this.props.locale === 'en-GB'} />
            <Header.Utility.Item id="en-US" text="en-US" key="en-US" isSelected={this.props.locale === 'en-US'} />
            <Header.Utility.Item id="de" text="de" key="de" isSelected={this.props.locale === 'de'} />
            <Header.Utility.Item id="es" text="es" key="es" isSelected={this.props.locale === 'es'} />
            <Header.Utility.Item id="fr" text="fr" key="fr" isSelected={this.props.locale === 'fr'} />
            <Header.Utility.Item id="pt" text="pt" key="pt" isSelected={this.props.locale === 'pt'} />
            <Header.Utility.Item id="fi-FI" text="fi-FI" key="fi-FI" isSelected={this.props.locale === 'fi-FI'} />
          </Header.Utility.ItemGroup>,
        ]}
      />
    );

    let themeSwitcher;

    function supportsCSSVars() {
      return window.CSS && window.CSS.supports && window.CSS.supports('(--fake-var: 0)');
    }

    if (supportsCSSVars()) {
      themeSwitcher = (
        <Header.Utility.Item
          text={`Theme: ${this.props.theme}`}
          key="theme"
          subMenuItems={[
            <Header.Utility.ItemGroup isSelectable key="theme-options" onChange={this.props.onThemeChange} >
              <Header.Utility.Item id="Default Theme" text="Default Theme" key="default" isSelected={this.props.theme === 'Default Theme'} />
              <Header.Utility.Item id="Consumer Theme" text="Consumer Theme" key="consumer" isSelected={this.props.theme === 'Consumer Theme'} />
              <Header.Utility.Item id="Mock Theme" text="Mock Theme" key="mock" isSelected={this.props.theme === 'Mock Theme'} />
            </Header.Utility.ItemGroup>,
          ]}
        />
      );
    } else {
      themeSwitcher = <div />;
    }

    const utility = (
      <Header.Utility
        accessory={<IconSettings />}
        title={'Config'}
        menuItems={[themeSwitcher, localeContent, <Header.Utility.Divider key="DIVIDER-1" />, bidiContent]}
      />
    );

    let navTabs;
    if (this.props.navigation && !isCompactHeader) {
      navTabs = <NavTabs links={this.props.navigation.links} />;
    }

    return (
      <Header
        layoutConfig={this.props.layoutConfig}
        logo={(
          <Header.Logo
            title="Terra"
            subtitle="Framework"
            accessory={<Image variant="rounded" src="https://github.com/cerner/terra-core/raw/master/terra.png" height="26px" width="26px" isFluid />}
          />
        )}
        utility={utility}
        content={navTabs}
      />
    );
  }
}

ApplicationHeader.propTypes = propTypes;

export default ApplicationHeader;
