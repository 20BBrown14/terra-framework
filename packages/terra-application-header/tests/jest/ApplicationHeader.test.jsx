import React from 'react';

import ApplicationHeader from '../../src/ApplicationHeader';

describe('ApplicationHeader', () => {
  it('should render without optional props', () => {
    const result = shallow((
      <ApplicationHeader />
    ));
    expect(result).toMatchSnapshot();
  });

  it('should render with provided logo and utility props', () => {
    const result = shallow((
      <ApplicationHeader
        logo={<ApplicationHeader.Logo />}
        utility={<ApplicationHeader.Utility />}
      />
    ));

    expect(result).toMatchSnapshot();
  });
});
