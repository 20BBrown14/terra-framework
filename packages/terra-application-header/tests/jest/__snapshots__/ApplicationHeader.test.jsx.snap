// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationHeader should render with provided logo and utility props 1`] = `
<div
  className="header fill"
>
  <div
    className="fill header-body"
  >
    <div
      className="fit start logo"
    >
      <Logo
        size="tiny"
      />
    </div>
    <div
      className="fill"
    />
    <div
      className="fit end utility"
    >
      <Utility
        size="tiny"
      />
    </div>
  </div>
</div>
`;

exports[`ApplicationHeader should render without optional props 1`] = `
<div
  className="header fill"
/>
`;
